src/main/java/com/ebivariation/contigalias/service/AssemblyService.java

```diff
@@ -16,36 +16,152 @@
+        Optional<AssemblyEntity> assembly = getAssemblyByAccession(accession);
+        if (assembly.isPresent()) {
+            return assembly;
+        }
+        Optional<AssemblyEntity> fetchAssembly = fetchAndInsertAssembly(accession);
+        if (fetchAssembly.isPresent()) {
+            AssemblyEntity asm = fetchAssembly.get();
             List<ChromosomeEntity> chromosomes = asm.getChromosomes();
             if (chromosomes != null) {
                 chromosomes.forEach(chr -> chr.setAssembly(null));
             }
-        });
+            // Setting this to null to remove inconsistency in results
+            // When fetching from db bring no results, chromosomes is an empty list
+            // whereas here chromosomes is value
+            else {
+                asm.setChromosomes(new LinkedList<>());
+            }
+            return fetchAssembly;
+        }
+        return Optional.empty();
+    }
+
+    public Optional<AssemblyEntity> fetchAndInsertAssembly(String accession) throws IOException {
+        if (repository.findAssemblyEntityByAccession(accession).isPresent()) {
+            throw duplicateAssemblyInsertionException(accession);
+        }
+        Optional<AssemblyEntity> fetchAssembly = dataSource.getAssemblyByAccession(accession);
+        fetchAssembly.ifPresent(this::insertAssembly);
+        return fetchAssembly;
+    }
+
+    public Optional<AssemblyEntity> getAssemblyByAccession(String accession) {
+        Optional<AssemblyEntity> assembly = repository.findAssemblyEntityByAccession(accession);
+        assembly.ifPresent(this::stripAssemblyFromChromosomes);
         return assembly;
     }
 
+    private void stripAssemblyFromChromosomes(AssemblyEntity assembly) {
+        List<ChromosomeEntity> chromosomes = assembly.getChromosomes();
+        if (chromosomes != null && chromosomes.size() > 0) {
+            chromosomes.forEach(chr -> chr.setAssembly(null));
+        }
+    }
+
+    public void insertAssembly(AssemblyEntity entity) {
+        setCacheSizeLimit();
+
+        if (isEntityPresent(entity)) {
+            throw duplicateAssemblyInsertionException(entity);
+        } else {
+            repository.save(entity);
+        }
+    }
+
+    /**
+     * Limits the size of the cache to a maximum of 10 assemblies
+     * <p>
+     * I'm using a while loop instead of an if statement because
+     * if two requests reach at once, they both might read cache
+     * size < 10 and add an entry leading to cache having more than
+     * 10 entries. While loop on next run deletes entities till cache
+     * size < 10. Now of course the same problem can arise if two
+     * requests start deleting at the same time but all that will lead
+     * to is the cache getting completely emptied.
+     * </p>
+     */
+    private void setCacheSizeLimit() {
+        // Limit cache size to 10 assemblies
+        while (repository.countByIdNotNull() >= 10) {
```

Comment:
the comment doesn't need to be there anymore and the number should be extracted to a constant. Replace the number with the constant in the javadoc too.